lib/compress.js

```diff
@@ -1749,8 +1749,14 @@ merge(Compressor.prototype, {
                     extract_declarations_from_unreachable_code(compressor, self.body, a);
                     return make_node(AST_BlockStatement, self, { body: a });
                 }
+            } else {
+                // self instanceof AST_Do
+                return self.body;
```

Comment:
`return self;` would certainly fix the do-while case. I've already confirmed this.